### PR TITLE
Allow addditional assert function names with `expect-expect`

### LINKED ISSUE #154: Ability to extend/override assertion function names
Related to #67.

`eslint-plugin-jest`'s `expect-expect` rule has a configuration option for overwriting the list of assertion function names. This is handy if you write functions that contain complex assertions reusable across several different tests. This plugin's `expect-expect` does not have this option, however.

If acceptable, I can try to open a PR to add this feature.

### PR DESCRIPTION
**Pull request by sarayourfriend**
Fixes #154.

This PR does not attempt parity with the `eslint-plugin-jest` rule option. It would not necessarily be difficult to achieve parity. The `jest` plugin is MIT licensed and we could copy/paste the various utility functions from that library to implement a similarly flexible option.

I've opted not to do that because this naive approach was very simple to implement and I wanted to get the PR up for it in case it's an acceptable MVP for this change. However, if y'all would like me to extend it to add full parity with the `jest` plugin's rule's options, I am happy to do so, just let me know.

### CODE REVIEW

**Comment by gilmeir-arnac** [comment]
+1 for customized function names for assertions enforcement, that's exactly what I was looking for 🙏

**Comment by sarayourfriend** [comment]
@mskelton Can you advise whether this approach is acceptable or if you'd prefer 1:1 parity with the jest plugin's rule configuration?

**Comment by mskelton** [comment]
Sorry, haven't had a chance for review. I think what I would actually prefer would be to use global settings so that we can share those settings for all rules that are checking expect statements. Similar to this:

https://github.com/jest-community/eslint-plugin-jest#aliased-jest-globals

**Comment by sarayourfriend** [comment]
Sounds good, I'll draft this and hope to return to work on it later this week.

**Comment by sarayourfriend** [comment]
@mskelton I've got a working implementation based on shared data. However, there are two problems I'd like to clarify:

- The expansion of valid assert functions really only applies to `expect-expect`. `valid-expect`, for example (the other obvious callsite of `isExpectCall`) only refers to actual usages of `expect`. Custom assertion functions couldn't be evaluated by `valid-expect` without heaps more configuration. In a sense `valid-expect` ensures that calls to `expect` actually create assertions. Allowing custom assert function names in `expect-expect` is a bit circuitous. A better overall rule might be a combination of both `expect-expect` and `valid-expect` that accepts custom assertion function names, named something like `expect-assertion`. `valid-expect` essentially _ensures the assertion_, while `expect-expect` just ensures that some call to `expect` is made, regardless of whether there is an assertion. I'm not suggesting a big refactor here to combine the rules or anything, just to suggest that I'm sceptical of the wider use case of configuring assertion functions in shared data _if the reason is to make them available to other rules that evaluate assertions_, specifically because there aren't other rules that a generic plugin could have that could apply to custom assertion functions.
- Using _only_ shared data to configure the list of additional assertion function names has some limitations, primarily that a module-specific assertion function would need to be configured in the global ESLint configuration rather than at the module level using the `/* eslint <rule name>: <options> */` pragma available for rule options. I'd suggest we should use _both_. It is very convenient to have a global list of additional assertion functions, especially one that persists even if specific modules need to expand the list (using just rule options means globally configured options would need to be repeated in modules if they're needed there, not ideal). But needing to define a reusable assertion function in the global list that is only relevant to a particular module is tedious and leaks module-specific details into the global configuration. Would you be open to the PR including _both_ shared data-based configuration and rule options-based configuration? This is predicated on me not having been able to find any way to modify shared data on a module-specific level. If that is possible, then forgive me ignorance and ignore this bit. It's also only relevant if you agree that the additional assertion function names is only applicable to `expect-expect` in the first place.

**Comment by sarayourfriend** [comment]
@mskelton Any thoughts regarding my question above? Are you okay with taking the dual approach, allowing for both adding assert function names that should be considered valid in all modules as well as the rule-specific configuration, to allow to extending the list in a given module?

**Comment by mskelton** [comment]
@sarayourfriend Sorry, I forgot to reply to this thread. I think I agree that doing both is probably best. Global config that can be overridden per rule.

**Comment by sarayourfriend** [comment]
No worries, thanks for confirming. I'll try to have the PR ready be the end of the week.

**Comment by sarayourfriend** [comment]
This PR should be ready for another review @mskelton :pray:

File: src/rules/expect-expect.ts [hunk 1]
```diff
@@ -1,10 +1,35 @@
...
+import { Settings } from '../utils/types';
+
+type Options = {
+  additionalAssertFunctionNames?: readonly string[];
+};
+
+function isAssertionCall(
+  node: ESTree.CallExpression,
+  additionalAssertFunctionNames: readonly string[]
+) {
```
Comment (mskelton) [inline src/rules/expect-expect.ts:13]: Lot's of `readonly` in this file, not really necessary

File: src/rules/expect-expect.ts [hunk 2]
```diff
@@ -1,10 +1,35 @@
...
+    )
+  );
+}
+
+function getAdditionalAssertFunctionNames(context: Rule.RuleContext): readonly string[] {
+  const globalSettings = (context.settings as Settings).playwright?.additionalAssertFunctionNames ?? [] as readonly string[]
+  const ruleSettings = (context.options[0] as Options | undefined)?.additionalAssertFunctionNames ?? [] as readonly string[]
+
+  return [...globalSettings, ...ruleSettings]
+}
```
Comment (mskelton) [inline src/rules/expect-expect.ts:27]: Maybe move this to a utils file for reuse

Comment (sarayourfriend) [reply src/rules/expect-expect.ts:27]: It's not necessarily useful for other rules, considering its dependency on the rule's own options (`context.options[0]` in this case). Likewise, the issues with re-using this logic elsewhere that I outlined in my previous PR comment. Happy to move it if you'd like, I'm just not sure that it's a coherent signal to contributors in the future about how applicable this particular function is to rules aside from this one. Are there particular use-cases you have in mind that I could add to a doc comment on the function if it is moved?

File: package.json [hunk 3]
```diff
@@ -19,7 +19,7 @@
     "lint": "eslint .",
     "format": "prettier --write .",
     "format:check": "prettier --check .",
-    "test": "jest",
+    "test": "pnpm run build && jest",
```
[the remark this answers is not in the document]
Comment (sarayourfriend) [reply package.json]: Sounds good. I'll open an issue to add a DEVELOPMENT.md to cover the correct way to run tests, etc.

File: src/rules/expect-expect.ts (hunk 1 above)
Comment (sarayourfriend) [reply src/rules/expect-expect.ts:13]: It doesn't create a restriction to the passer, just consumer. I can remove it if you like, but it's useful type information for both perspectives (function writer knows/is disallowed from mutation, caller knows mutation won't happen), in my experience.

Comment (mskelton) [reply src/rules/expect-expect.ts:13]: It's unnecessarily verbose, so please remove it.

File: src/rules/expect-expect.ts (hunk 2 above)
Comment (mskelton) [reply src/rules/expect-expect.ts:27]: There are other rules that would benefit from the global settings, so yes let's move it out.

**Comment by sarayourfriend** [comment]
@mskelton Your requests are addressed in the most recent commit.

File: src/utils/misc.ts [hunk 4]
```diff
@@ -1,4 +1,19 @@
+import { Rule } from 'eslint';
+
+import { Settings } from './types';
+
 export const getAmountData = (amount: number) => ({
   amount: amount.toString(),
   s: amount === 1 ? '' : 's',
 });
+
+type AssertFunctionNamesOptions = {
```
Comment (mskelton) [inline src/utils/misc.ts:10]:
```suggestion
interface AssertFunctionNamesOptions {
```

**Review by mskelton — APPROVED** [review]

**Pull request merged**